### ISSUE TITLE
Quick question on attentionSum model

### ISSUE THREAD

**Issue by spookyQubit** [zhanghaoyu1993/RC-experiments#4]
Hi,
I had a question which I cannot seem to figure out by reading the paper. This is regarding calculation of score before optimization. 

Suppose we have the following case:

Document:
Obama and Trump are going to meet. Elon will not meet him. 

Question:
Who will meet Obama?

Answer candidates:
{Trump, Elon}

True answer:
Trump

Because we know that the answer has to lie within the answer candidates, do we calculate the score (which will get converted to prob and ultimately be used for optimization) only for "Trump" and "Elon"? Thant is, do we calculate scores only for the candidates: {s(Trump), s(Elon)} ? 

Or the score is calculated for each and every word in the document? That is, do we calculate {s(Obama), s(and), s(Trump), s(are), s(going) ... s(him), s(.)} ? 

Would really appreciate your input on this. 
Thanks.

**Comment by cairoHy** [comment]
That depends on the model, on attentionSumReader, the attention is calculated on every context word, as the sentence below(from the paper) .
>>> To form a proper probability distribution over the words in the document, we normalize the weights using the softmax function.

**Comment by spookyQubit** [comment]
Thanks for the quick reply :+1: 

That is what I got from reading the paper too. But does it not then mean that they are ignoring the candidate words all together?

**Comment by cairoHy** [comment]
No, during attention sum process, the candidate words will be used.

**Comment by spookyQubit** [comment]
Thanks again for the reply. 

Yes, I read the paper again and as you correctly describe, the probability is calculated over all words. 

So, from what I understand right now, I have
During training:
1) Calculate the scores (then probabilities) for each word in the document
2) Calculate the probability of the answer (requires sum over wherever the answer occurs in the document)
2) Minimize -log(prob(answer))

During prediction:
1) Calculate the score (then probability) for each word in the document (This is same as training)
2) Calculate the probability of all candidate answers (requires sum over wherever the candidate occurs in the document)
3) Predicted answer = Candidate with max probability

Note that in my current understanding, during training, the probability of only the true answer needs to be calculated. On the other hand, during prediction we need to calculate the probability of each candidate answer. Is this understanding correct? 

Thanks a lot again. 

P.S.  I am trying to implement the model in: https://github.com/spookyQubit/dynetAttentionSumReader/blob/master/ASReaderTrainer.py#L45

**Comment by cairoHy** [comment]
During training, the prob of all candidate word are calculated and calculate the cross entropy of pred prob and true answer(one-hot form).

**Comment by spookyQubit** [comment]
Thanks a lot again for your replies.
